src/mscorlib/src/System/Threading/ExecutionContext.cs

```diff
@@ -274,14 +274,29 @@ internal static void SetLocalValue(IAsyncLocal local, object newValue, bool need
         {
             ExecutionContext current = Thread.CurrentThread.ExecutionContext ?? ExecutionContext.Default;
 
+            // No change, fast exit
+            if (newValue == null && current == ExecutionContext.Default)
+                return;
+
             object previousValue;
             bool hadPreviousValue = current.m_localValues.TryGetValue(local, out previousValue);
 
+            // No change, fast exit
             if (previousValue == newValue)
                 return;
 
             IAsyncLocalValueMap newValues = current.m_localValues.Set(local, newValue);
 
+            if (newValue == null && newValues.Count == 0)
+            {
+                // Async locals cleared - set ExecutionContext back to Default to move back to fast-path
+                Thread.CurrentThread.ExecutionContext = ExecutionContext.Default;
```

Comment:
ExecutionContext contains more than just the async locals. Do you also need to compare the rest of the state?